qiskit/pulse/parser.py

```diff
@@ -0,0 +1,292 @@
+        self._params.add(node.id)
+        return node
+
+    def visit_UnaryOp(self, node):
+        """Evaluate unary operation and return ast.Num if operand is bound.
+
+        Args:
+            node (ast.UnaryOp): Unary operation to evaluate.
+
+        Returns:
+            ast.UnaryOp or ast.Num: Evaluated value.
+        """
+        node.operand = self.visit(node.operand)
+        if isinstance(node.operand, ast.Num):
+            val = ast.Num(n=self._match_ops(node.op, self._unary_ops,
+                                            node.operand.n))
+            return ast.copy_location(val, node)
+        return node
+
+    def visit_BinOp(self, node):
+        """Evaluate binary operation and return ast.Num if operands are bound.
+
+        Args:
+            node (ast.BinOp): Binary operation to evaluate.
+
+        Returns:
+            ast.BinOp or ast.Num: Evaluated value.
+        """
+        node.left = self.visit(node.left)
+        node.right = self.visit(node.right)
+        if isinstance(node.left, ast.Num) and isinstance(node.right, ast.Num):
+            val = ast.Num(n=self._match_ops(node.op, self._binary_ops,
+                                            node.left.n, node.right.n))
+            return ast.copy_location(val, node)
+        return node
+
+    def visit_Call(self, node):
+        """Evaluate function and return ast.Num if all arguments are bound.
+
+        Args:
+            node (ast.Call): Function to evaluate.
+
+        Returns:
+            ast.Call or ast.Num: Evaluated value.
+
+        Raises:
+            PulseError: When unsupported or unsafe function is specified.
+        """
+        if not isinstance(node.func, ast.Name):
+            raise PulseError('Unsafe expression is detected.')
+        node.args = [self.visit(arg) for arg in node.args]
+        if all(isinstance(arg, ast.Num) for arg in node.args):
+            if node.func.id not in self._math_ops.keys():
+                raise PulseError('Function %s is not supported.' % node.func.id)
+            _args = [arg.n for arg in node.args]
+            _val = self._math_ops[node.func.id](*_args)
+            if not _val.imag:
+                _val = _val.real
+            val = ast.Num(n=_val)
+            return ast.copy_location(val, node)
+        return node
+
+    def generic_visit(self, node):
+        raise PulseError('Unsupported node: %s' % node.__class__.__name__)
+
+
+def parse_string_expr(source, partial_binding=False):
+    """Safe parsing of string expression.
+
+    Args:
+        source (str): String expression to parse.
+        partial_binding (bool): Allow partial bind of parameters.
+
+    Returns:
+        PulseExpression: Returns a expression object.
+    """
+    subs = [('numpy.', ''), ('np.', ''), ('math.', ''), ('cmath.', '')]
+    for match, sub in subs:
+        source = source.replace(match, sub)
+
+    expression = PulseExpression(source, partial_binding)
+
+    return expression, expression.params
```

Comment:
If `params` are already an attribute of `expression` maybe we should only just return `expression`?

Reply:
I agree. This is just to align with old interface. I'll update converter module to use params property.